### ISSUE TITLE
h1 element in Accessibility Quiz - RWD

### ISSUE THREAD

**Issue by Supravisor** [freeCodeCamp/freeCodeCamp#58213]
### Describe the Issue

1) In the Full Stack Developer beta curriculum, the suggestion for accessibility is to include content after the heading element.

https://www.freecodecamp.org/learn/full-stack-developer/lecture-importance-of-accessibility-and-good-html-structure/how-does-proper-heading-level-structure-affect-accessibility

[video timestamp](https://youtu.be/FqBrK9IAvhI?t=140): 2.20

![Image](https://github.com/user-attachments/assets/3a4ff041-acb9-4da1-af4c-d2708c219c56)
![Image](https://github.com/user-attachments/assets/012ac931-0a1b-4357-89b5-5e9cab8e42a3)

A suggestion is to include text after the `<h1>` element which describes the purpose of the "A11y" quiz.

<hr />

2) Also, to raise what was incorrectly placed elsewhere:

For both accessibility and search engine optimisation `<h1>` elements are either child elements of the `<body>` or `<header>` elements, not `<main>` elements

* [SEO best practices for h1 element: H1 headers are located in the `body` of the webpage, while title tags go in the `head` section. ](https://ahrefs.com/blog/h1-tag/#:~:text=H1%20headers%20are%20located%20in,title%20tags%20as%20SERP%20headlines)
* [W3C page structure example code](https://www.w3.org/WAI/tutorials/page-structure/example/)
* [Ensures that the page, or at least one of its frames, contains an h1 element that appears before the start of the main content to allow screen reader users to use keyboard shortcuts to navigate the heading structure instead of wasting time listening to more of the web page to understand its structure.](https://dequeuniversity.com/rules/axe/4.1/page-has-heading-one)
* [Use one unique h1 per page that describes what that page is about. That h1 preferably starts just above the main content.](https://www.a11yproject.com/posts/how-to-accessible-heading-structure/)
* [H69: Providing heading elements at the beginning of each section of content Example 3](https://www.w3.org/TR/WCAG20-TECHS/H69)

There are a wide range of views regarding where to place the `<h1>` element, which I found out when researching the list above.  However, most of the documentation supports placing the `<h1>` element before the main content.

At the moment, the order is `<header>`, which contains the `<h1>` element, then the `<main>` element
[PR #57944](https://github.com/freeCodeCamp/freeCodeCamp/pull/57944) places the `<h1>` element as the first child of the `<main>` element.

I have no issue with the order of the `<header>` element or where it is currently placed.  It makes sense to announce site navigation for screen reader users before other content.

### Affected Page

https://www.freecodecamp.org/learn/2022/responsive-web-design/learn-accessibility-by-building-a-quiz/step-7

### Your code

```



```

### Expected behavior
Edit
Proposed:
* after the navigation container, nest the the `<h1>` element  in an `<header>` element and include a short descriptive text about the Accessibility Quiz before the `<main>` element.

### System

- Device: [e.g. iPhone 6, Laptop]
- OS: [e.g. iOS 14, Windows 10, Ubuntu 20.04]
- Browser: [e.g. Chrome, Safari]
- Version: [e.g. 22]

**Comment by bbsmooth** [comment]
I'll try to clarify the best practices for the h1:

- There should only be one h1 on a page
- It should generally be near the top of the page
- I should reflect the main purpose of the page (is often similar to the title of the page)
- It should be within a landmark. The two most common options are the header element (which creates the banner landmark) or the main element (which creates the main landmark)

There are certainly arguments for placing it in the header, but the overwhelming choice is to put it at the beginning of the main element. This is by far the most common placement and thus screen reader users expect it to be there.

As for the list of five links provided to support not placing the h1 inside the main element:

- The article states that "H1 headers are located in the `<body>` of the webpage", but this is meaningless since all page content is within in the body element.
- The second link is a reputable example, they have just chosen to put the h1 in the header. As I said, there are arguments for putting it in the header, but there are even more arguments for putting it in the main element. Regardless, this is just an example, it isn't telling you that the h1 should always be in the header element.
- The third link is from a reputable accessibility organization and is describing a rule in the Axe core accessibility checker. The rule description states "Ensures that the page, or at least one of its frames, contains an h1 element that appears before the start of the main content" which is correct. But notice they said "main content", not "main element". The main content will be within the main element (if the page is coded accessibly), so that implies that the h1 will also be within the main element. This rule is really just checking if there is an h1 on the page. It's not checking where that h1 is located. There is another Axe rule which checks that [all content is within a landmark](https://dequeuniversity.com/rules/axe/4.1/region). This is the rule that will be flagged if your h1 isn't inside either the header or main element.
- Same thing as the previous link ("main content" vs. "main element").
- The fifth link points to an old version of this document. If you look at the [most recent version](https://www.w3.org/WAI/WCAG22/Techniques/html/H69), you'll see that the examples are using the main element and have the h1 heading inside the main element.

**Comment by Supravisor** [comment]
1. The first article was making a distinction between the `h1` element and `title` element. The article was included to emphasise search engine optimisation.
2. From [https://www.w3.org/WAI/tutorials/page-structure/](https://www.w3.org/WAI/tutorials/page-structure/):
>Well-structured content allows more efficient navigation and processing. Use HTML and WAI-ARIA to improve navigation and orientation on web pages and in applications.
...
These tutorials provide best-practice guidance on implementing accessibility in different situations.

>Regardless, this is just an example, it isn't telling you that the h1 should always be in the header element.

The tutorials and code example are providing guidance on best practices, though do not stipulate any standards.
3. See note below
4. [Best practices summarized](https://www.a11yproject.com/posts/how-to-accessible-heading-structure/)
>Use one unique h1 per page that describes what that page is about. That h1 preferably starts just above the main content.
5. The [newer H69 document](https://www.w3.org/WAI/WCAG22/Techniques/html/H69) did not show up in the search results, so was not accessible to my browser. Maybe due to how the page was coded, maybe not:
```
<main id="main" class="standalone-resource__main">
<h1><span>Technique H69:</span>Providing heading elements at the beginning of each section of content</h1>
```

<hr />

The "main" difference between the Accessibility practice project and [the W3C example](https://www.w3.org/WAI/tutorials/page-structure/example/), the `h1` element is not a child element of `main`:

```
<body>
  <header>
    <h1>Space Teddy Inc.</h1>
  </header>
  <nav aria-labelledby="mainnavheader">
    <h2 id="mainnavheader">Navigation Menu</h2>
    …
  </nav>
  <main>
    <article>
      <h2>An inside look at the new Space Teddy 6</h2>
 ```

<hr />

Note
I edited the "Expected Behaviour" section to nest the `h1` element in a `header` element as you mentioned either `header` or `main` as landmark alternatives.

**Comment by Supravisor** [comment]
I listed the W3C example because they gave an implementation of code structure that supports an A11y Project best practice  - one `h1` element above the main content.

When I visit the page I can see code for example two:

![Image](https://github.com/user-attachments/assets/d859900e-d41c-4924-be12-fd6771c89072)

A layout will depend on several factors such as sidebars, navigational containers, banners, and viewport size.
So creating a rigid rule for `h1`,`header`, and `main` elements may not be practical and as you mentioned there can be various use cases.

<hr />

The Deque University page titled ["All page content must be contained by landmarks"](https://dequeuniversity.com/rules/axe/4.1/region) with HTML5 or ARIA landmark regions which you refer to shows two versions of markup.  The rule description is:

>It is best practice to contain all content excepting skip links, within distinct regions such as the header, nav, main, and footer.

Using that structure as a guide to design a web page, it could make sense to replace the text "This is the header" with an `h1` or ARIA banner element in the `header` element.

According to an [MDN article for the banner role](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/banner_role) both of the code examples do not nest the `h1` element in a `main` element.

ARIA role element:
```
<div role="banner">
  <a href="#main" id="skipToMain" class="skiptocontent">Skip To main content</a>
  <img src="images/w3c.png" alt="W3C Logo" />
  <h1>ARIA Landmarks</h1>
  <p>Identifying page subsections for easy navigation</p>
  <nav>…</nav>
</div>
```

Semantic header element:
```
<header>
  <a href="#main" id="skipToMain" class="skiptocontent">Skip To main content</a>
  <img src="images/w3c.png" alt="W3C Logo" />
  <h1>ARIA Landmarks</h1>
  <p>Identifying page subsections for easy navigation</p>
  <nav>…</nav>
</header>
```

<hr />

I agree with @bbsmooth that:
>The h1 should be within either the header or main element.

<hr />

>Also, notice the wording here: site name vs content heading. h1 can be inside the main element if it's used as the content heading (or to use the verbiage in this thread, the heading of the main content).

Could this mean no accompanying text for the `h1` element?

<hr />

I did not refer to W3C, the A11y Project, or Deque University as "reputable sources".

**Comment by bbsmooth** [comment]
If I understand correctly, this issue is now suggesting two changes to the Accessibility Quiz project:

1. Move the h1 immediately after the nav element in the header
2. Add some short descriptive text about the quiz before the main element

I don't understand why 1 is necessary. The h1 is visually placed before the navigation, so it should also be before the navigation in the DOM. I do not agree that this proposed change should be made.

I'm not opposed to 2, I don't understand the need for it. And if we were going to add this text, it should be in the main landmark, not before it. Or are you suggesting it would be in the header element?

**Comment by Supravisor** [comment]
1. Issue #47830 originally raised the issue of:
>Move the h1 element out of the header and make it the first element in main. Best practices would say that the h1 element should be the first element in main (but this is not an accessibility requirement).

Could not find any "Best practices" supporting the idea of placing the `h1` element as the first child of the `main` element.

Not adverse to taking the `h1` element out of the navigational element:
>I have no issue with the order of the `<header>` element or where it is currently placed. It makes sense to announce site navigation for screen reader users before other content.

Mentioned in an earlier post:
>A layout will depend on several factors such as sidebars, navigational containers, banners, and viewport size.
So creating a rigid rule for `h1`,`header`, and `main` elements may not be practical and as you mentioned there can be various use cases.

Did raise an issue for a new `header` element to contain the `h1` element.

Placing the `h1`  element before the main content is an A11y best practice according to the [A11y Project](https://www.a11yproject.com/posts/how-to-accessible-heading-structure/) and other sources.

Please refer to Issue #47830 regarding the latest PR.

2. Could the content of the main element count as the content belonging to the `h1` element?:
>Also, notice the wording here: site name vs content heading. h1 can be inside the main element if it's used as the content heading (or to use the verbiage in this thread, the heading of the main content).

I'm not sure what the above quote is suggesting which is why I asked the question about text accompanying the `h1` element.

>1) In the Full Stack Developer beta curriculum, the suggestion for accessibility is to include content after the heading element.

>![Image](https://github.com/user-attachments/assets/3a4ff041-acb9-4da1-af4c-d2708c219c56)
![Image](https://github.com/user-attachments/assets/012ac931-0a1b-4357-89b5-5e9cab8e42a3)

Does a lower level heading and its content count as content for a higher level heading?

If it does, then does the wording from [https://www.freecodecamp.org/learn/full-stack-developer/lecture-importance-of-accessibility-and-good-html-structure/how-does-proper-heading-level-structure-affect-accessibility
](https://www.freecodecamp.org/learn/full-stack-developer/lecture-importance-of-accessibility-and-good-html-structure/how-does-proper-heading-level-structure-affect-accessibility) need updating to reflect that?

>2. Add some short descriptive text about the quiz before the main element
 
>I'm not opposed to 2, I just don't understand the need for it. And if we were going to add this text, it should be in the main landmark, not before it. Or are you suggesting it would be in the header element?

Refer to the expected behaviour for the current issue and point 1 above regarding a new `header` element.

Headings can be a way to provide a few words regarding the content that follows it.
It aids in navigation and accessibility for web users, whether they are sighted, sight-impaired, or experience cognitive issues when accessing content.

Without good headings, it can make it much more difficult to access content.  Imagine having to go through a few paragraphs of text to discover that is not  content you wanted.  Sighted users could scan the text, others may find it necessary to consume a block of text that they then find they did not need to consume. I think good headings and their structure are important for accessibility, navigation, and for ensuring as many web users as possible, regardless of their abilities, can enjoy a good user experience.

In this instance I think descriptive text maybe necessary - the contents of the `h1` element  "HTML/CSS Quiz".  This is the freeCodeCamp Accessibility Quiz practice project, why not showcase A11y?

Placing the `h1` element inside the `main` element did affect the accessibility of a certain page to my search engine.
Without the direct link posted on this issue thread I could not access the web page.

**Comment by bbsmooth** [comment]
As I just mentioned in [#47830](https://github.com/freeCodeCamp/freeCodeCamp/issues/47830), I am officially withdrawing my recommendation to move the h1. The `h1` is fine where it is right now.

> Does a lower level heading and its content count as content for a higher level heading?

Yes.

> Could not find any "Best practices" supporting the idea of placing the h1 element as the first child of the main element.

> Placing the h1 element before the main content is an A11y best practice according to the [A11y Project](https://www.a11yproject.com/posts/how-to-accessible-heading-structure/) and other sources.

Ya, probably shouldn't have said that the `h1` should be the first element in `main` but rather that it should be placed before the main content on the page (just like the a11y Project recommends). And since the main content should be in the `main` element, that implies that the h1 will generally be at the beginning of the `main` element since all content, including the `h1` should be inside a landmark. 

Again, I'm not opposed to moving the `h1`, but if it is moved it must be placed within the `main` element.
